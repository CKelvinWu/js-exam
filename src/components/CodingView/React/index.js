import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import 'brace';
import 'brace/mode/jsx';
import 'brace/mode/javascript';
import 'brace/theme/textmate';
import 'brace/theme/monokai';

import Grid from 'components/Grid';
import GridItem from 'components/Grid/GridItem';
import ConsoleWidget from 'components/Widgets/ConsoleWidget';
import CodeWidget from 'components/Widgets/CodeWidget';
import ResultWidget from 'components/Widgets/ResultWidget';
import AnswerWidget from 'components/Widgets/AnswerWidget';

import debouncedRunCode from 'utils/runCode';
import { REACT as GRID_LABEL_REACT } from 'utils/gridLabel';

import styles from './ReactPage.module.scss';

class ReactPage extends Component {
  static propTypes = {
    isExaming: PropTypes.bool,
    code: PropTypes.string,
    compiledCode: PropTypes.string,
    consoleMsg: PropTypes.array,
    handleCodeChange: PropTypes.func,
    wrappedConsole: PropTypes.object,
    resetConsole: PropTypes.func,
  };

  controlHeight = 70;

  componentDidMount() {
    const { compiledCode, wrappedConsole, resetConsole } = this.props;
    resetConsole();
    debouncedRunCode({ code: compiledCode, wrappedConsole });
  }

  shouldComponentUpdate(nextProps) {
    const { compiledCode: previousCompiledCode } = this.props;
    const { compiledCode, wrappedConsole, resetConsole } = nextProps;
    if (previousCompiledCode !== compiledCode) {
      resetConsole();
      // WARNING: This is not debounced
      debouncedRunCode({ code: compiledCode, wrappedConsole });
    }
    return true;
  }

  render() {
    const { isExaming, code, handleCodeChange, consoleMsg } = this.props;
    const layout = [
      {
        key: 'code',
        x: 0,
        y: 0,
        width: window.innerWidth / 2 + 50,
        height: window.innerHeight / 2 + 100,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
      {
        key: 'answer',
        x: 0,
        y: 1,
        width: window.innerWidth / 2 + 50,
        height: window.innerHeight / 2 - 150,
        minWidth: 100,
        maxWidth: 700,
      },
      {
        key: 'result',
        x: 1,
        y: 0,
        width: window.innerWidth / 2 - 50,
        height: (window.innerHeight - this.controlHeight) / 2,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
      {
        key: 'console',
        x: 1,
        y: 1,
        width: window.innerWidth / 2 - 50,
        height: (window.innerHeight - this.controlHeight) / 2 - 50,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
    ];
    return (
      <div className={styles.app}>
        <Grid layout={layout} totalWidth="100%" totalHeight="100%" autoResize>
          <GridItem key="code" label={GRID_LABEL_REACT.code}>
            <CodeWidget
              handleCodeChange={handleCodeChange}
              data={code}
              readOnly={!isExaming}
              mode="jsx"
              theme="monokai"
            />
          </GridItem>
          <GridItem key="answer" label={GRID_LABEL_REACT.answer}>
            {isExaming ? <AnswerWidget /> : null}
          </GridItem>
          <GridItem key="result" label={GRID_LABEL_REACT.result}>
            {isExaming ? <ResultWidget /> : null}
          </GridItem>
          <GridItem key="console" label={GRID_LABEL_REACT.console}>
            {isExaming ? <ConsoleWidget data={consoleMsg} /> : null}
          </GridItem>
        </Grid>
      </div>
    );
  }
}

export default withRouter(ReactPage);
