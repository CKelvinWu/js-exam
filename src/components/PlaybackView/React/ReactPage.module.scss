.app {
  display: flex;
  flex-direction: row;
}

.rightWidget {
  height: 33.3%;
  width: 100%;
  border: 1px solid #eee;
}

.label {
	position: absolute;
  margin-top: 5px;
  right: 17px;
	z-index: 10;
	padding: 2px 12px;
	color: white;
	background-color: rgba($color: #000000, $alpha: 0.5);
}