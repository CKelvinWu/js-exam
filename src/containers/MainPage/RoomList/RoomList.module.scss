.list {
  list-style-type: none;
}

.listItem {
  border-radius: 10px;

  .tag {
    position: absolute;
    top: 5px;
    left: 5px;
    z-index: 1;
  }
}

.cover {
  padding: 10px;
  font-size: 25px;
  color: #000;
  text-align: center;
  background-color: #eee;
  position: relative;

  .room {
    display: block;
    color: #333;
    font-size: 16px;
  }
  .floatTop {
    position: absolute;
    right: 3%;
    top: 5%;
    font-size: 16px;
  }
  .floatTop:hover {
    transition: all 0.1s ease-in-out;
    transform: scale(1.3);
  }
}
