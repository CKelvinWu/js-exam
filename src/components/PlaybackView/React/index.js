import React, { Component } from 'react';
import PropTypes from 'prop-types';
import 'brace';
import 'brace/mode/jsx';
import 'brace/mode/javascript';
import 'brace/theme/textmate';
import 'brace/theme/monokai';

import { Spin } from 'antd';

import CodeWidget from 'components/Widgets/CodeWidget';
import ResultWidget from 'components/Widgets/ResultWidget';
import AnswerWidget from 'components/Widgets/AnswerWidget';

import debouncedRunCode from 'utils/runCode';
import { REACT as GRID_LABEL_REACT } from 'utils/gridLabel';

import styles from './ReactPage.module.scss';

class ReactPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      height: props.height,
      widthCode: props.width,
      widthTests: props.width,
    };
  }

  async componentDidMount() {
    const { compiledCode, addTape } = this.props;

    this.setState({
      height: `${window.innerHeight - 160}px`,
      widthCode: `${window.innerWidth * 0.45}px`,
      widthTests: `${window.innerWidth * 0.3}px`,
    });
    debouncedRunCode({ code: compiledCode, onTapeUpdate: addTape });
  }

  shouldComponentUpdate(nextProps) {
    
    const { compiledCode: previousCompiledCode } = this.props;
    const { compiledCode } = nextProps;
    if (previousCompiledCode !== compiledCode) {
      debouncedRunCode({ code: compiledCode });
    }
    return true;
  }

  render() {
    this.codeWidgetstyle = {
      height: this.state.height,
      width: this.state.widthCode,
      position: 'relative',
    };
    this.testWidgetStyle = {
      height: this.state.height,
      width: this.state.widthTests,
      position: 'relative',
    };
    const { handleCodeChange, test, code } = this.props;
    return (
      <React.Fragment>
        <div className={styles.app}>
          <div style={this.codeWidgetstyle}>
            <div className={styles.label}>Code</div>
            <CodeWidget
              handleCodeChange={handleCodeChange}
              data={code}
              mode="jsx"
              theme="monokai"
              readOnly
            />
          </div>
          <div style={this.testWidgetStyle}>
            <div className={styles.rightWidget}>
              <div className={styles.label}>{GRID_LABEL_REACT.test}</div>
              <CodeWidget data={test} mode="jsx" theme="textmate" readOnly />
            </div>
            <div className={styles.rightWidget}>
              <div className={styles.label}>{GRID_LABEL_REACT.answer}</div>
              <AnswerWidget />
            </div>
            <div className={styles.rightWidget}>
              <div className={styles.label}>{GRID_LABEL_REACT.result}</div>
              <ResultWidget />
            </div>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

ReactPage.propTypes = {
  compiledCode: PropTypes.string,
  code: PropTypes.string,
  height: PropTypes.string,
  width: PropTypes.string,
  test: PropTypes.string,
  handleCodeChange: PropTypes.func,
  addTape: PropTypes.func,
};

export default ReactPage;
