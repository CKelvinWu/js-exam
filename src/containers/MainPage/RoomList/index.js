import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import { List, Avatar } from 'antd';
import style from './RoomList.module.scss';

const RoomList = ({ rooms, isLoading, signedOn, hostings }) => (
  <List
    itemLayout="horizontal"
    dataSource={rooms}
    loading={isLoading}
    renderItem={room => (
      <Link
        to={{
          pathname: `/admin/dispatch/${room.id}`,
        }}
      >
        <List.Item
          style={{ borderBottom: '1px solid #ddd' }}
          className={style.listItem}
        >
          <List.Item.Meta
            className={style.listMeta}
            avatar={<Avatar icon="home" className={style.avatar} />}
            title={
              <div>
                <div className={style.roomNameHeader}>Room </div>
                <div className={style.roomName}>{room.description}</div>
              </div>
            }
            description={
              <div>
                <div className={style.subjectIdHeader}>Candidate </div>
                <div className={style.subjectId}>{room.subjectId}</div>
              </div>
            }
          />
          {signedOn && hostings.includes(room.id) && (
            <div className={style.listContent}>Host</div>
          )}
        </List.Item>
      </Link>
    )}
  />
);

RoomList.propTypes = {
  rooms: PropTypes.array,
  isLoading: PropTypes.bool,
  signedOn: PropTypes.bool,
  hostings: PropTypes.array,
};

export default RoomList;
