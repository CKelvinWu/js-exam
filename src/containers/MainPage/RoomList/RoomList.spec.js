import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';

import RoomList from './index';

const fakeProps = {
  rooms: [
    {
      id: 'f509ee49-c4d6-4d3c-b357-2d18454b01f2',
      subjectId: 'test room 1',
      description: 'Z77',
      createTime: '2020-01-31T07:07:03.295Z',
      createTimeByDate: new Date(
        'Fri Jan 31 2020 15:07:03 GMT+0800 (Taipei Standard Time)',
      ),
    },
    {
      id: 'c6f86538-3661-4cd3-a8e2-09578958bd39',
      subjectId: 'test room 2',
      description: 'V72',
      createTime: '2020-01-31T07:06:41.936Z',
      createTimeByDate: new Date(
        'Fri Jan 31 2020 15:06:41 GMT+0800 (Taipei Standard Time)',
      ),
    },
  ],
  signedOn: true,
  hostings: ['f509ee49-c4d6-4d3c-b357-2d18454b01f2'],
  triggerDelRoom: jest.fn(),
};

describe('<RoomList />', () => {
  let subject;

  const setup = props =>
    mount(
      <BrowserRouter>
        <RoomList {...props} />
      </BrowserRouter>,
    );

  beforeEach(() => {
    jest.resetModules();
    jest.resetAllMocks();
    subject = setup(fakeProps);
  });

  it('should render', () => {
    expect(subject.length).toBeTruthy();
  });

  it('should render as snapshot', () => {
    expect(subject.html()).toMatchSnapshot();
    expect(subject.find('li').length).toEqual(2);
    expect(subject.find('button i.anticon-delete').length).toEqual(1);
  });

  it('should render properly when no data', () => {
    const props = Object.assign({}, fakeProps);
    delete props.rooms;
    subject = setup(props);
    expect(subject.html()).toMatchSnapshot();
  });

  it('should trigger triggerDelRoom function when click the delete button', () => {
    const delButton = subject.find('button i.anticon-delete');
    delButton.simulate('click');
    expect(subject.prop('children').props.triggerDelRoom).toBeCalledWith(
      subject.prop('children').props.rooms[0],
    );
  });
});
