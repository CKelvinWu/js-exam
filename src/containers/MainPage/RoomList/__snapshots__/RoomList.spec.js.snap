// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomList /> should render as snapshot 1`] = `"<span><ul style=\\"position: relative; width: 1095px; height: 200px;\\" class=\\"list\\"><li class=\\"listItem\\" style=\\"position: absolute; top: 0px; left: 0px; z-index: 2; opacity: 1; transform: translateX(0px) translateY(0px) scale(1);\\"><span class=\\"ant-tag ant-tag-has-color tag\\" style=\\"background-color: rgb(16, 142, 233);\\">Host</span><a href=\\"/admin/dispatch/f509ee49-c4d6-4d3c-b357-2d18454b01f2\\"><div class=\\"ant-card ant-card-bordered ant-card-hoverable\\"><div class=\\"ant-card-cover\\"><div class=\\"cover\\"><span class=\\"room\\">Room</span>Z77<button type=\\"button\\" class=\\"floatTop\\"><i aria-label=\\"icon: delete\\" class=\\"anticon anticon-delete\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"delete\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path fill=\\"#ffeae6\\" d=\\"M292.7 840h438.6l24.2-512h-487z\\"></path><path fill=\\"#f00\\" d=\\"M864 256H736v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zm-504-72h304v72H360v-72zm371.3 656H292.7l-24.2-512h487l-24.2 512z\\"></path></svg></i></button></div></div><div class=\\"ant-card-body\\"><div class=\\"ant-card-meta\\"><div class=\\"ant-card-meta-avatar\\"><span class=\\"ant-avatar ant-avatar-circle ant-avatar-image\\"><img src=\\"avatar3.png\\"></span></div><div class=\\"ant-card-meta-detail\\"><div class=\\"ant-card-meta-title\\">test room 1</div><div class=\\"ant-card-meta-description\\">2020/01/31 15:07:03</div></div></div></div></div></a></li><li class=\\"listItem\\" style=\\"position: absolute; top: 0px; left: 0px; z-index: 2; opacity: 1; transform: translateX(275px) translateY(0px) scale(1);\\"><a href=\\"/admin/dispatch/c6f86538-3661-4cd3-a8e2-09578958bd39\\"><div class=\\"ant-card ant-card-bordered ant-card-hoverable\\"><div class=\\"ant-card-cover\\"><div class=\\"cover\\"><span class=\\"room\\">Room</span>V72</div></div><div class=\\"ant-card-body\\"><div class=\\"ant-card-meta\\"><div class=\\"ant-card-meta-avatar\\"><span class=\\"ant-avatar ant-avatar-circle ant-avatar-image\\"><img src=\\"avatar1.png\\"></span></div><div class=\\"ant-card-meta-detail\\"><div class=\\"ant-card-meta-title\\">test room 2</div><div class=\\"ant-card-meta-description\\">2020/01/31 15:06:41</div></div></div></div></div></a></li></ul></span>"`;

exports[`<RoomList /> should render properly when no data 1`] = `"<span></span>"`;
